File: sol/2290/2290.cpp
#include <queue>

struct Solution {
    int minimumObstacles(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        vector<vector<bool>> reached(m, vector<bool>(n, false));
        deque<pair<int, pair<int, int>>> q;
        q.push_back({0, {0, 0}});
        while (!q.empty()){
            int cost = q.back().first;
            auto [i, j] = q.back().second;
            q.pop_back();
            if (i == m - 1 && j == n - 1){
                return cost;
            }
            if (reached[i][j]){
                continue;
            }
            reached[i][j] = true;
            int ii = i + 1, jj = j;
            if (ii < m && !reached[ii][jj]){
                if (grid[ii][jj]){
                    q.push_front({cost + 1, {ii, jj}});
                } else {
                    q.push_back({cost, {ii, jj}});
                }
            }
            ii = i, jj = j + 1;
            if (jj < n && !reached[ii][jj]){
                if (grid[ii][jj]){
                    q.push_front({cost + 1, {ii, jj}});
                } else {
                    q.push_back({cost, {ii, jj}});
                }
            }
            ii = i - 1, jj = j;
            if (ii >= 0 && !reached[ii][jj]){
                if (grid[ii][jj]){
                    q.push_front({cost + 1, {ii, jj}});
                } else {
                    q.push_back({cost, {ii, jj}});
                }
            }
            ii = i, jj = j - 1;
            if (jj >= 0 && !reached[ii][jj]){
                if (grid[ii][jj]){
                    q.push_front({cost + 1, {ii, jj}});
                } else {
                    q.push_back({cost, {ii, jj}});
                }
            }
        }
        return -1;
    }
};

File: sol/2290/2290.py
from collections import deque

class Solution:
    def minimumObstacles(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        reached = [[False] * n for _ in range(m)]
        q = deque([(0, 0, 0)])
        while len(q) != 0:
            cost, i, j = q.pop()
            if i == m - 1 and j == n - 1:
                return cost
            if reached[i][j]:
                continue
            reached[i][j] = True
            for ii, jj in ((i + 1, j), (i, j + 1), (i - 1, j), (i, j - 1)):
                if not (0 <= ii < m and 0 <= jj < n) or reached[ii][jj]:
                    continue
                if grid[ii][jj]:
                    q.appendleft((cost + 1, ii, jj))
                else:
                    q.append((cost, ii, jj))
        return -1

